File: bin/day-13.py
import re
from collections import defaultdict

import time
from itertools import permutations

parser = re.compile(r'(\w+) would (gain|lose) (\d+) happiness units by sitting next to (\w+)')

prefs = defaultdict(dict)

for line in open("../inputs/day-13.txt"):
    parts = parser.match(line).groups()
    val = int(parts[2]);
    if parts[1] == 'lose':
        val *= -1
    prefs[parts[0]][parts[3]] = val

# part 2
for person in prefs.keys():
    prefs['Andre'][person] = 0
    prefs[person]['Andre'] = 0

best_change = 0
now = time.time()
for perm in permutations(prefs.keys(), len(prefs.keys())):
    change = 0
    for i, person in enumerate(perm):
        change += prefs[person][perm[i-1]]
        change += prefs[person][perm[i+1 if i < len(perm)-1 else 0]]
    if change > best_change:
        best_change = change

print best_change, "in", (time.time() - now) * 1000, 'ms'

File: bin/day-13.dart
import 'dart:io';

RegExp parser = new RegExp(r'(\w+) would (gain|lose) (\d+) happiness units by sitting next to (\w+)');

Iterable<List> permutations(Iterable src) sync* {
  if (src.length == 1) {
    yield src;
  } else {
    for (var item in src) {
      for (List p in permutations(src.where((i) => i != item).toList())) {
        yield p..add(item);
      }
    }
  }
}

main() {
  Map<String, Map<String, int>> prefs = {};
  Match m;
  int val;
  for (String line in new File("./inputs/day-13.txt").readAsLinesSync()) {
    m = parser.firstMatch(line);
    if (!prefs.containsKey(m.group(1))) {
      prefs[m.group(1)] = {};
    }
    val = int.parse(m.group(3));
    if (m.group(2) == 'lose') {
      val *= -1;
    }
    prefs[m.group(1)][m.group(4)] = val;
  }
  // Part 2
  prefs['Andre'] = {};
  for (String k in prefs.keys) {
    prefs['Andre'][k] = 0;
    prefs[k]['Andre'] = 0;
  }

  int best = -9999999;
  int permLen = prefs.keys.length;
  Stopwatch watch = new Stopwatch()
    ..start();
  for (List<String> perm in permutations(prefs.keys)) {
    int sum = 0;
    for (int i = 0; i < permLen; i++) {
      sum += prefs[perm[i]][perm[i == 0 ? permLen - 1 : i - 1]];
      sum += prefs[perm[i]][perm[i == permLen - 1 ? 0 : i + 1]];
    }
    if (sum > best) {
      best = sum;
    }
  }
  watch.stop();
  print("$best took ${watch.elapsedMilliseconds}ms");
}